### ISSUE TITLE
Reference for IRK3 integrator

### ISSUE THREAD

**Issue by ctowery** [spectralDNS/shenfun#73]
Hi,

I was just curious about what particular kind of RK scheme was being used in the IRK3 integrator, and if there was a particular reference from which you drew the coefficients?

Thanks.

**Comment by mikaem** [comment]
Hi
I believe it was drawn from the Simson KTH solver. But it's been a while, would be great if you could check it out😀
Mikael

**Comment by ctowery** [comment]
The scheme is indeed published (without reference) in the 2007 SIMSON technical report from KTH. The overall scheme is semi-implicit, paring an explicit Low-Storage RK33 (3-stage, 3rd-order) for nonlinear terms with Crank-Nicolson for linear terms, so rather than being an "IRK" or "SIRK" scheme, wherein any implicit part is baked into the RK, it would get labeled (by a method paper for RK schemes) something like RK(3,3)-2N + CN, where the 2N refers to the kind of low-storage scheme it seems to be. 

I worked backwards to some early SIMSON method papers but never saw an explicit reference for the exact LSRK Butcher coefficients used, but at this point any modern LSRK reference, such as those pointed to by Nodepy, describe how to construct an RK(3,3)-2N scheme, so the only open question was why those particular coefficients were chosen over other examples. 

The original scheme paper for all 2N LSRK schemes seems to be Williamson 1980.

Thanks for the help!

**Comment by mikaem** [comment]
Thanks a lot for the clarification! I have spent far too little time on the temporal integrators. Any upgrade or additions would be greatly appreciated😀

**Comment by ctowery** [comment]
I found a lead in Appendix D of Canuto, Quarteroni, Hussaini, and Zang (new editions, Vol. 2 on CFD methods), where they reference a Spalart and Watmuff 1993 JFM publication as providing "another common LSRK+CN" method due to Wray (unpublished), but in fact I finally found the SIMSON RK(3,3)-2N coefficients in a 1991 JCP paper from Spalart, Moser, and Rogers (https://doi.org/10.1016/0021-9991(91)90238-G), in which they say the coefficients are only "almost 3rd order", so the original Williamson 1980 (3,3) coefficients or the (4,3) coefficients from Carpenter and Kennedy (1994) might provide better results just by swapping coefficients (and bumping the rk loops by 1 stage), but obviously the best person to ask about optimal coefficients would be an integrator expert like David Ketcheson. I'm just a turbulence modeling/theory guy 😃.

Anyway, that concludes the hunt for where the coefficients came from.

I am curious if you'd ever tested any of the IRK3/ETD/ETDRK4 integrators from shenfun using a CFL-based adaptive time step, and thus had to call the setup() function for the solver over and over?  Would this be crazy inefficient, or would the alltoall cost of a large-scale 3D HIT or channel flow make the solver matrix-assembly steps seem utterly cheap in comparison?  If you haven't ever tried then I'll definitely run some tests to see what I could get away with.  I'm running decaying HIT cases using a slight modification of your shenfun_NS example, basically the spectralLES module I developed as a PhD student (https://github.com/ctowery/teslapy/tree/master/spectralLES) but upgraded to shenfun, and since the maximum time step increases dramatically over time, there will be a time interval where re-running the setup for a bigger dt will be worth it. (Question is if that interval is closer to every step or 1000 steps.)
